modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/RemoveWriteOnGcInvokeClosure.java

```diff
@@ -0,0 +1,180 @@
+
+    private OperationType operationType;
+
+    private @Nullable VersionChain newRow;
+
+    private List<RowVersion> toRemove;
+
+    private RowVersion result;
+
+    private @Nullable RowVersion toUpdate;
+
+    RemoveWriteOnGcInvokeClosure(RowId rowId, HybridTimestamp timestamp, AbstractPageMemoryMvPartitionStorage storage) {
+        this.rowId = rowId;
+        this.timestamp = timestamp;
+        this.storage = storage;
+    }
+
+    @Override
+    public void call(@Nullable VersionChain oldRow) throws IgniteInternalCheckedException {
+        assert oldRow != null : "rowId=" + rowId + ", storage=" + storage.createStorageInfo();
+        assert oldRow.hasNextLink() : oldRow;
+
+        RowVersion rowVersion = findRowVersionLinkWithChecks(oldRow);
+        RowVersion nextRowVersion = storage.readRowVersion(rowVersion.nextLink(), ALWAYS_LOAD_VALUE, true);
+
+        result = nextRowVersion;
+
+        // If the found version is a tombstone, then we must remove it as well.
+        if (rowVersion.isTombstone()) {
+            toRemove = List.of(nextRowVersion, rowVersion);
+
+            // If the found version is the head of the chain, then delete the entire chain.
+            if (oldRow.headLink() == rowVersion.link()) {
+                operationType = OperationType.REMOVE;
+            } else if (oldRow.nextLink() == rowVersion.link()) {
+                operationType = OperationType.PUT;
+
+                // Find the version for which this version is RowVersion#nextLink.
+                toUpdate = storage.readRowVersion(oldRow.headLink(), ALWAYS_LOAD_VALUE, false);
+
+                newRow = oldRow.setNextLink(nextRowVersion.nextLink());
+            } else {
+                operationType = OperationType.PUT;
+
+                newRow = oldRow;
+
+                // Find the version for which this version is RowVersion#nextLink.
+                toUpdate = storage.foundRowVersion(oldRow, equalsByNextLink(rowVersion.link()), false);
+            }
+        } else {
+            operationType = OperationType.PUT;
+
+            toRemove = List.of(nextRowVersion);
+
+            toUpdate = rowVersion;
+
+            if (oldRow.headLink() == rowVersion.link()) {
+                newRow = oldRow.setNextLink(nextRowVersion.nextLink());
+            } else {
+                newRow = oldRow;
+            }
+        }
+    }
+
+    @Override
+    public @Nullable VersionChain newRow() {
+        assert operationType == OperationType.PUT ? newRow != null : newRow == null : "newRow=" + newRow + ", op=" + operationType;
+
+        return newRow;
+    }
+
+    @Override
+    public OperationType operationType() {
+        assert operationType != null;
+
+        return operationType;
+    }
+
+    @Override
+    public void onUpdate() {
+        if (toUpdate != null) {
+            try {
+                storage.rowVersionFreeList.updateNextLink(toUpdate.link(), result.nextLink());
```

Comment:
Do you really have to do this while holding a write lock on a tree page?
This is interesting actually. Technically, no RO transaction should attempt reads below LW, so entry that's being deleted should never be read. Please think carefully about this place.

Reply:
I think what should be left here, when filling the cursor cache in `AbstractPartitionTimestampCursor`, we can get an error when trying to read the removed version.